flatflow/scheduler/internal/algorithm/regression.h

```diff
@@ -33,62 +33,143 @@ namespace algorithm {
-template <std::size_t degree>
+template <std::size_t Order>
 class PassiveAggressiveRegressor {
  public:
-  inline explicit PassiveAggressiveRegressor(const double &epsilon)
+  inline explicit PassiveAggressiveRegressor(double epsilon)
       : epsilon_(epsilon) {
-    coef_ = std::vector<double>(degree + 1, 1.0);
-    power_ = std::vector<double>(degree + 1, 0.0);
-    coef_[0] = 0.0;
+    coef_ = std::vector<double>(Order, 1.0);
+    power_ = std::vector<double>(Order + 1, 0.0);
   }
 
   // PassiveAggressiveRegressor::fit()
   //
-  // Fits the model to the given vector data X and y.
+  // Fits the model to the given vector workloads and runtimes.
   // This implementation is based on Epsilon-Insensitive Passive Aggressive
-  // Regressor. This method is implementation of partialfit method in order to
-  // use in online learning.
-  inline void fit(const std::vector<double> &X, const std::vector<double> &y) {
-    CHECK_EQ(X.size(), y.size());
-    for (std::size_t iter = 0; iter < kMaxIter_; ++iter) {
-      for (std::size_t dataIdx = 0; dataIdx < X.size(); ++dataIdx) {
-        const auto prediction = predict(X[dataIdx]);
+  // Regressor. Unlike Polynomial regressor, current fit is used in
+  // heterogeneous environment.
+  inline bool fit(const std::vector<double> &workloads,
+                  const std::vector<double> &runtimes) {
+    CHECK_EQ(workloads.size(), runtimes.size());
+    bool early_stop = true;
+    for (std::size_t iter = 0; iter < max_iter_; ++iter) {
+      for (std::size_t index = 0; index < workloads.size(); ++index) {
+        const auto &workloard = workloads[index];
+        const auto &runtime = runtimes[index];
+        const auto prediction = predict(workloard);
         const auto loss =
-            std::max(0.0, std::abs(y[dataIdx] - prediction) - epsilon_);
+            std::max(0.0, std::abs(runtime - prediction) - epsilon_);
+        if (loss != 0.0) {
+          early_stop = false;
+          power_[0] = std::pow(workloard, 1);
+          power_[1] = std::pow(workloard, 2.0);
 
-        for (std::size_t coefIdx = 0; coefIdx < coef_.size(); ++coefIdx) {
-          power_[coefIdx] = std::pow(X[dataIdx], std::pow(2.0, coefIdx));
+          const auto xqnorm =
+              std::accumulate(power_.begin(), power_.end(), 0.0);
+          const auto update =
+              (std::min(C_, loss / xqnorm)) * ((runtime > prediction) ? 1 : -1);
+
+          intercept += update;
+          coef_[0] += update * workloard;
         }
+      }
+    }
+    return early_stop;
+  }
+  // PassiveAggressiveRegressor::predict()
+  //
+  // Predicts value based on given scalar worload and current model coef_.
+  inline const double predict(double workload) const {
+    return coef_[0] * workload;
+  }
+
+  double intercept = 0.0;
+
+ protected:
+  double epsilon_;
+  std::vector<double> coef_;
+  std::vector<double> power_;
+  static constexpr auto C_ = 1.0;
+  static constexpr auto max_iter_ = 1000;
```

Comment:
The type of ``max_iter_`` should be ``std::size_t``.
That is, ``static constexpr auto max_iter_ = static_cast<std::size_t>(1000);``.